### PR TITLE
Update the TF discussion of slice assignments and in-memory operations

### PR DESCRIPTION
**Pull request by tomerk**
Update the TF discussion of slice assignments and in-memory operations. Notes that Tensors are immutable, gradients do not flow backwards through variable writes, and that putting computations in a `tf.function` can minimize the number of allocations tensorflow needs to do.

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_preliminaries/ndarray.md [hunk 1]
```diff
@@ -545,15 +594,29 @@ print('id(z):', id(z))
...
+  z = tf.zeros_like(y)  # This unused value will be pruned out.
+  a = x + y  # Allocations will be re-used when no longer needed.
+  b = a + y
+  c = b + y
+  return c + y
+
+computation(x, y)
 ```
 
+:begin_tab:`mxnet`
```
Comment (terrytangyuan) [inline chapter_preliminaries/ndarray.md:608]: You can use `mxnet,pytorch` here for shared tab

File: chapter_preliminaries/ndarray.md [hunk 2]
```diff
@@ -451,7 +451,23 @@ selects the second and the third elements as follows:
 x[-1], x[1:3]
 ```
 
+:begin_tab:`mxnet`
 Beyond reading, we can also write elements of a matrix by specifying indices.
+:end_tab:
+
+:begin_tab:`pytorch`
```
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:458]: We can combine tabs with `mxnet, pytorch`

File: chapter_preliminaries/ndarray.md [hunk 3]
```diff
@@ -514,6 +530,8 @@ allocating memory unnecessarily all the time.
 In machine learning, we might have
 hundreds of megabytes of parameters
 and update all of them multiple times per second.
+
+:begin_tab:`mxnet`
```
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:534]: same here

File: chapter_preliminaries/ndarray.md (hunk 1 above)
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:608]: same here

**Comment by mli** [comment]
Job d2l-en/PR-1103/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1103/

**Comment by mli** [comment]
Job d2l-en/PR-1103/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1103/

File: chapter_preliminaries/ndarray.md [hunk 4]
```diff
@@ -514,19 +526,39 @@ allocating memory unnecessarily all the time.
 In machine learning, we might have
 hundreds of megabytes of parameters
 and update all of them multiple times per second.
+
```
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:529]: I'll remove this newline in my fix. No action needed.

File: chapter_preliminaries/ndarray.md [hunk 5]
```diff
@@ -514,19 +526,39 @@ allocating memory unnecessarily all the time.
...
 
+:begin_tab:`mxnet`
 Fortunately, performing in-place operations in MXNet is easy.
+:end_tab:
+
+:begin_tab:`pytorch`
+Fortunately, performing in-place operations in PyTorch is easy.
+:end_tab:
+
+:begin_tab:`mxnet, pytorch`
```
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:544]: Will fix these separated paragraphs in my fix. no action needed

**Comment by astonzhang** [comment]
Thanks Tomer!

**Comment by astonzhang** [comment]
FYI, my minor fix: https://github.com/d2l-ai/d2l-en/commit/1f3d1a5a5b3f30259fdcf19a2b87d510dd01da3c
